modules/luci-base/root/usr/libexec/generate_otp.uc

```diff
@@ -0,0 +1,340 @@
+		let a = h0;
+		let b = h1;
+		let c = h2;
+		let d = h3;
+		let e = h4;
+
+		for (j = 0; j < 80; j++) {
+			let f = 0;
+			let k = 0;
+
+			// Setup reange constants
+			if (j < 20) {
+				f = (b & c) | ((~b) & d);
+				k = 0x5A827999;
+			} else if (j < 40) {
+				f = b ^ c ^ d;
+				k = 0x6ED9EBA1;
+			} else if (j < 60) {
+				f = (b & c) | (b & d) | (c & d);
+				k = 0x8F1BBCDC;
+			} else {
+				f = b ^ c ^ d;
+				k = 0xCA62C1D6;
+			}
+
+			let temp = circular_shift(a, 5) + f + e + k + block[j] & 0xFFFFFFFF;
+                        e = d;
+                        d = c;
+                        c = circular_shift(b, 30);
+                        b = a;
+                        a = temp;
+		}
+
+		// Update hash values
+		h0 = (h0 + a) & 0xFFFFFFFF;
+                h1 = (h1 + b) & 0xFFFFFFFF;
+                h2 = (h2 + c) & 0xFFFFFFFF;
+                h3 = (h3 + d) & 0xFFFFFFFF;
+                h4 = (h4 + e) & 0xFFFFFFFF;
+	}
+
+	let sha1 = [];
+
+	h0_binary = intToBin(h0);
+	for (i = 0; i < length(h0_binary); i++)
+		sha1[i] = h0_binary[i];
+
+	h1_binary = intToBin(h1);
+	for (i = 0; i < length(h1_binary); i++)
+		sha1[i+4] = h1_binary[i];
+
+	h2_binary = intToBin(h2);
+	for (i = 0; i < length(h2_binary); i++)
+		sha1[i+8] = h2_binary[i];
+
+	h3_binary = intToBin(h3);
+	for (i = 0; i < length(h3_binary); i++)
+		sha1[i+12] = h3_binary[i];
+
+	h4_binary = intToBin(h4);
+	for (i = 0; i < length(h4_binary); i++)
+		sha1[i+16] = h4_binary[i];
+
+	return sha1;
+}
+
+function calculate_hmac_sha1(key, message) {
+	const message_binary = strToBin(message);
+	let binary_key = strToBin(key);
+
+	if (length(key) > 64)
+		binary_key = calculate_sha1(binary_key);
+
+	for (i = 0; i < 64 - length(key); i++)
+		binary_key[length(key)+i] = 0x0;
+
+	let ko = [];
+	for (i = 0; i < 64; i++)
+		ko[i] = binary_key[i] ^ 0x36;
+
+	for (i = 0; i < length(message); i++)
+		ko[64+i] = message_binary[i];
+
+	const sha1_ko = calculate_sha1(ko);
+
+	ko = [];
+
+	for (i = 0; i < 64; i++)
+		ko[i] = binary_key[i] ^ 0x5c;
+
+	for (i = 0; i < length(sha1_ko); i++)
+		ko[64+i] = sha1_ko[i];
+
+	const hmac = calculate_sha1(ko);
+
+	return hmac;
+}
+
+function calculate_hotp(key, counter)
+{
+	const secret = encode_base32(key);
+	const counter_bytes = [ 0x0, 0x0, 0x0, 0x0,
+				(counter >> 24) & 0xff,
+				(counter >> 16) & 0xff,
+				(counter >> 8) & 0xff,
+				counter & 0xff ];
+
+	const digest = calculate_hmac_sha1(secret, binToStr(counter_bytes));
+
+	const offset_bits = digest[19] & 0xf;
+
+	let p = [];
+	for (i = 0; i < 4; i++)
+		p[i] = digest[offset_bits+i];
+
+	const snum = (p[0] << 24 | p[1] << 16 | p[2] << 8 | p[3]) & 0x7fffffff;
+	const otp = snum % 10 ** 6;
+
+	return otp;
+}
+
+function get_otp(username)
+{
+	const ctx = cursor();
+
+	let key = ctx.get('luci', username, 'key');
+	if (!key) {
+		printf("Missing key for user %s\n", username);
+		exit(1);
+	}
+
+	let otp_type = ctx.get('luci', username, 'type');
+	let counter = "";
```

Comment:
I don't know how ucode responds to 64bit uints 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt

64bit may be necessary to avoid Y2038 problems.

Reply:
It consistently uses 64bit integers internally.

Reply:
@jow- even on 32bit targets? Soo %d is actually long int?

Reply:
Yes, even on 32bit targets.

Reply:
I think %d is just `int` - irrespective of length. a 32bit platform probably does twice as many operations and then glues the lower and upper halves together again.